android/walletlib/src/main/java/com/solana/mobilewalletadapter/walletlib/scenario/LocalScenario.java

```diff
@@ -0,0 +1,403 @@
+                            @NonNull Scenario.Callbacks callbacks,
+                            @NonNull byte[] associationPublicKey) {
+        this(context, mobileWalletAdapterConfig, authIssuerConfig, callbacks,
+                associationPublicKey, new DevicePowerConfigProvider(context));
+    }
+
+    /*package*/ LocalScenario(@NonNull Context context,
+                              @NonNull MobileWalletAdapterConfig mobileWalletAdapterConfig,
+                              @NonNull AuthIssuerConfig authIssuerConfig,
+                              @NonNull Scenario.Callbacks callbacks,
+                              @NonNull byte[] associationPublicKey,
+                              @NonNull PowerConfigProvider powerConfigProvider) {
+        mCallbacks = callbacks;
+        mMobileWalletAdapterConfig = mobileWalletAdapterConfig;
+        this.associationPublicKey = associationPublicKey;
+
+        final LooperThread t = new LooperThread();
+        t.start();
+        mIoLooper = t.getLooper(); // blocks until Looper is available
+        mIoHandler = new Handler(mIoLooper);
+
+        mAuthRepository = new AuthRepositoryImpl(context, authIssuerConfig);
+
+        this.mPowerManager = powerConfigProvider;
+    }
+
+    @Override
+    public byte[] getAssociationPublicKey() {
+        return associationPublicKey;
+    }
+
+    @Override
+    protected void finalize() {
+        mIoLooper.quitSafely();
+    }
+
+    public MessageReceiver createMessageReceiver() {
+        return new MobileWalletAdapterSession(
+                this,
+                new MobileWalletAdapterServer(mMobileWalletAdapterConfig, mIoLooper, mMethodHandlers),
+                mSessionStateCallbacks);
+    }
+
+    @Override
+    public void start() {
+        startNoConnectionTimer();
+    }
+
+    @Override
+    public abstract void close();
+
+    public Long getNoConnectionTimeout() {
+        return mPowerManager.isLowPowerMode() ? mMobileWalletAdapterConfig.noConnectionWarningTimeoutMs : 0L;
+    }
+
+    private void startNoConnectionTimer() {
+        // we cant actually check if a connection is still alive, so instead we start a timer
+        // that assumes we do not have connection if the timeout is reached. Therefore, we
+        // MUST cancel this timer if we receive any connections or messages before it ends
+        Long noConnectionTimeout = getNoConnectionTimeout();
+        if (noConnectionTimeout > 0)
+            mNoConnectionTimeoutHandler = mTimeoutExecutorService.schedule(() -> {
+                Log.i(TAG, "No connection timeout reached");
+                if (mCallbacks instanceof Callbacks) {
+                    ((Callbacks) mCallbacks).onLowPowerAndNoConnection();
+                }
+            }, noConnectionTimeout, TimeUnit.MILLISECONDS);
+    }
+
+    private void stopNoConnectionTimer() {
+        if (mNoConnectionTimeoutHandler != null) mNoConnectionTimeoutHandler.cancel(true);
```

Comment:
Should we also null out the `mNoConnectionTimeoutHandler` here, so it can be garbage collected?